test/distributed/checkpoint/test_torch_save_to_dcp.py

```diff
@@ -0,0 +1,248 @@
+        ).redistribute(
+            placements=placement,
+        )
+
+    return state_dict
+
+
+def _offline_convert(fp_name: str, target_dir: str, world_size: int):
+    # For offline conversion, we only need to create a state_dict with
+    # FQN->DTensor mapping and save it with DCP.save().
+    state_dict = _construct_sharded_state_dict(fp_name, world_size)
+    DCP.save(state_dict, checkpoint_id=target_dir)
+
+
+def _construct_state_dict_and_metadata(
+    fp_name: str, world_size: int
+) -> Tuple[Dict[str, Any], Metadata]:
+    """
+    This step depends on the source checkpoint format. In our case, the
+    source checkpoint format is torch.save(), we can only do torch.load() to
+    construct the state_dict.
+    """
+
+    # Not much we can do to save the memory because of the way we save the
+    # source checkpoint. For other checkpoint formats, one can do incremental
+    # load if possible to avoid CPU OOM.
+    state_dict = torch.load(fp_name, map_location="cpu")
+
+    state_dict_metadata = {}
+    placement = [Shard(0)]
+    for key, value in state_dict.items():
+        # Sharded the tensor into a DTensor with each rank contains a shard.
+        size = value.size()
+        state_dict_metadata[key] = TensorStorageMetadata(
+            TensorProperties(dtype=value.dtype),
+            size,
+            [ChunkStorageMetadata(size, torch.Size([0 for _ in size]))],
+        )
+    return state_dict, Metadata(state_dict_metadata=state_dict_metadata)
+
+
+class TorchSaveReader(StorageReader):
+    def __init__(self, fp_name: str, world_size: int) -> None:
+        # This is likely to cause CPU OOM if the model size is large.
+        # An alternative solution is to actually load the data in # `read_data()`.
+        # The performance may be slow depending on the source checkpointing format.
+        state_dict, metadata = _construct_state_dict_and_metadata(fp_name, world_size)
+        super().__init__()
+        self.state_dict = state_dict
+        self.metadata = metadata
+
+    def read_data(self, plan: LoadPlan, planner: LoadPlanner) -> Future[None]:
+        for req in plan.items:
+            if req.type == LoadItemType.BYTE_IO:
+                raise RuntimeError("The conversion does not support BYTE_IO yet.")
+            else:
+                tensor = self.state_dict[req.storage_index.fqn]
+                tensor = narrow_tensor_by_index(
+                    tensor, req.storage_offsets, req.lengths
+                )
+                target_tensor = planner.resolve_tensor(req).detach()
+                assert target_tensor.size() == tensor.size(), (
+                    f"req {req.storage_index} mismatch sizes, "
+                    f"{target_tensor.size()} vs {tensor.size()}"
+                )
+                target_tensor.copy_(tensor)
+                planner.commit_tensor(req, target_tensor)
+
+        fut: Future = Future()
+        fut.set_result(None)
```

Comment:
strange use of future?

Reply:
This is because `read_data` requires a future. So we simply create one and assign a None to it. I'm not sure why the original intention to have a `Future[None]` as the return value.